editor/src/components/editor/actions/actions.spec.browser2.tsx

```diff
@@ -1171,6 +1176,87 @@ describe('actions', () => {
         )
       })
     })
+
+    describe('end-to-end copy paste', () => {
+      const clipboardMock = new MockClipboardHandlers()
+      beforeEach(() => {
+        clipboardMock.setup()
+      })
+
+      afterEach(() => {
+        clipboardMock.teardown()
+      })
+
+      it('can copy-paste end-to-end', async () => {
+        const testCode = `
+        <div data-uid='aaa' style={{contain: 'layout', width: 300, height: 300}}>
+          <div data-uid='bbb'>
+            <div data-uid='ccc' style={{position: 'absolute', left: 20, top: 50, bottom: 150, width: 100}} />
+            <div data-uid='ddd' style={{width: 60, height: 60}} />
+          </div>
+        </div>
+      `
+        const renderResult = await renderTestEditorWithCode(
+          makeTestProjectCodeWithSnippet(testCode),
+          'await-first-dom-report',
+        )
+        await selectComponentsForTest(renderResult, [makeTargetPath('aaa/bbb')])
+        await pressKey('c', { modifiers: cmdModifier })
+
+        await selectComponentsForTest(renderResult, [makeTargetPath('aaa')])
+
+        const canvasRoot = renderResult.renderedDOM.getByTestId('canvas-root')
+
+        FOR_TESTS_setNextGeneratedUids(['child1', 'child2', 'parent'])
+
+        firePasteEvent(canvasRoot)
+
+        // Wait for the next frame
+        await clipboardMock.pasteDone
```

Comment:
I think it'd be more ergonomic to make this the return value of `firePasteEvent` and use it as
```ts
await firePasteEvent(clipboardMock, canvasRoot)
```

(you could introduce another helper that ties the two together but imho it'd be redundant)

Even if there were cases where you don't want to `await` right after the call, you could always save the promise and use it independently

Reply:
The promise is here because Clipboard.parseClipboardData is originally an async function (since it might read files from the clipboard, which is an async operation), and the pasteDone promise basically exposes this fact. It's not ideal that an implementation detail (Clipboard.parseClipboardData being async) is exposed, but since one is already knowingly mocking the clipboard functions this isn't that big a deal